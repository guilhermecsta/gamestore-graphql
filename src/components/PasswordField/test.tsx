import { render, screen, waitFor } from 'utils/test-utils'
import userEvent from '@testing-library/user-event'

import PasswordField from '.'

describe('<PasswordField />', () => {
  it('should show visibility toggle only when at least one letter is typed', async () => {
    render(<PasswordField label="PasswordField" />)

    const btnToggleShowPassword = screen.getByRole('img', {
      name: /show password/i
    })

    expect(btnToggleShowPassword).not.toBeVisible()

    const input = screen.getByTestId(/password-input/i)
    const text = 'This is my new text'
    userEvent.type(input, text)

    expect(btnToggleShowPassword).toBeVisible()

    userEvent.click(btnToggleShowPassword)
    await waitFor(() => {
      expect(
        screen.getByRole('img', { name: /hide password/i })
      ).toBeInTheDocument()
    })
  })

  it('should render with label', () => {
    render(<PasswordField label="label" name="Label" />)

    expect(screen.getByLabelText('label')).toBeInTheDocument()
  })

  it('should render without label', () => {
    render(<PasswordField />)

    expect(screen.queryByLabelText('Label')).not.toBeInTheDocument()
  })

  it('should render with placeholder', () => {
    render(<PasswordField placeholder="hey you" />)

    expect(screen.getByPlaceholderText('hey you')).toBeInTheDocument()
  })

  it('should change its value when typing', async () => {
    const onInputChange = jest.fn()
    render(
      <PasswordField
        onInputChange={onInputChange}
        label="PasswordField"
        name="PasswordField"
      />
    )

    const input = screen.getByTestId(/password-input/i)
    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).toHaveValue(text)
      expect(onInputChange).toHaveBeenCalledTimes(text.length)
    })
    expect(onInputChange).toHaveBeenCalledWith(text)
  })

  it('should be accesible using tab', () => {
    render(<PasswordField label="PasswordField" name="PasswordField" />)

    const input = screen.getByLabelText('PasswordField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).toHaveFocus()
  })

  it('should not change its value when disabled', async () => {
    const onInputChange = jest.fn()
    render(
      <PasswordField
        onInputChange={onInputChange}
        label="PasswordField"
        name="PasswordField"
        disabled
      />
    )

    const input = screen.getByTestId(/password-input/i)
    expect(input).toBeDisabled()

    const text = 'This is my new text'
    userEvent.type(input, text)

    await waitFor(() => {
      expect(input).not.toHaveValue(text)
    })
    expect(onInputChange).not.toHaveBeenCalled()
  })

  it('should not be accesible using tab when disabled', () => {
    render(
      <PasswordField label="PasswordField" name="PasswordField" disabled />
    )

    const input = screen.getByLabelText('PasswordField')
    expect(document.body).toHaveFocus()

    userEvent.tab()
    expect(input).not.toHaveFocus()
  })

  it('should render with error', () => {
    const { container } = render(
      <PasswordField
        label="PasswordField"
        name="PasswordField"
        error="Error message"
      />
    )

    expect(screen.getByText('Error message')).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render with loading', () => {
    render(
      <PasswordField
        label="PasswordField"
        name="PasswordField"
        loading="Validating..."
      />
    )

    expect(screen.getByText('Validating...')).toBeInTheDocument()
  })
})
