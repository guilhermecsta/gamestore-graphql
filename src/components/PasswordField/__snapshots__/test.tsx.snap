// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PasswordField /> should render with error 1`] = `
.c7 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #EAEAEA;
  border-radius: 0.2rem;
  padding: 0 1.6rem;
  border: 0.2rem solid;
  border-color: #EAEAEA;
}

.c4:focus-within {
  box-shadow: 0 0 0.5rem #F231A5;
}

.c8 {
  color: #030517;
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
  padding: 0.8rem 0;
  padding-left: 2.4rem;
  background: transparent;
  border: 0;
  outline: none;
  width: 100%;
}

.c8:-webkit-autofill {
  -webkit-box-shadow: 0 0 0 2.4rem #EAEAEA inset;
  -webkit-filter: none;
  filter: none;
}

.c8:-webkit-autofill::first-line {
  font-family: Poppins,-apple-system,BlinkMacSystemFont,'Segoe UI',Roboto,Oxygen,Ubuntu,Cantarell,'Open Sans','Helvetica Neue',sans-serif;
  font-size: 1.6rem;
}

.c2 {
  font-size: 1.4rem;
  color: #030517;
  cursor: pointer;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  color: #8F8F8F;
  -webkit-order: 0;
  -ms-flex-order: 0;
  order: 0;
}

.c6 > svg {
  width: 2.2rem;
  max-height: 100%;
}

.c11 {
  color: #FF6347;
  font-size: 1.2rem;
}

.c11::before {
  content: '\\d7';
  font-size: 1.8rem;
  padding-right: 0.4rem;
  vertical-align: middle;
}

.c0 .c3 {
  border-color: #FF6347;
}

.c0 .c5,
.c0 .c1 {
  color: #FF6347;
}

.c9 {
  padding-left: 2.4rem;
  padding-right: 2.4rem;
  width: calc(100% - 4.4rem);
  -webkit-letter-spacing: 0.17rem;
  -moz-letter-spacing: 0.17rem;
  -ms-letter-spacing: 0.17rem;
  letter-spacing: 0.17rem;
  font-weight: 600;
}

.c9::-webkit-input-placeholder {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

.c9::-moz-placeholder {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

.c9:-ms-input-placeholder {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

.c9::placeholder {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

.c10 {
  -webkit-order: 1;
  -ms-flex-order: 1;
  order: 1;
  color: #b7b7b7;
  cursor: pointer;
  opacity: 1;
  -webkit-transition: opacity 0.3s ease-in-out;
  transition: opacity 0.3s ease-in-out;
  opacity: 0;
  pointer-events: none;
}

.c10:hover {
  color: #8F8F8F;
  -webkit-transition: 0.1s ease-in-out;
  transition: 0.1s ease-in-out;
}

<div
  class="c0"
>
  <label
    class="c1 c2"
    for="PasswordField"
  >
    PasswordField
  </label>
  <div
    class="c3 c4"
  >
    <div
      class="c5 c6"
    >
      <svg
        aria-hidden="true"
        class="c7"
        fill="currentColor"
        focusable="false"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
        >
          <path
            d="M0 0h24v24H0V0z"
          />
          <path
            d="M0 0h24v24H0V0z"
            opacity="0.87"
          />
        </g>
        <path
          d="M18 8h-1V6c0-2.76-2.24-5-5-5S7 3.24 7 6v2H6c-1.1 0-2 .9-2 2v10c0 1.1.9 2 2 2h12c1.1 0 2-.9 2-2V10c0-1.1-.9-2-2-2zM9 6c0-1.66 1.34-3 3-3s3 1.34 3 3v2H9V6zm9 14H6V10h12v10zm-6-3c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2z"
        />
      </svg>
    </div>
    <input
      aria-busy="false"
      aria-invalid="true"
      aria-live="polite"
      class="c8 c9"
      data-testid="password-input"
      id="PasswordField"
      name="PasswordField"
      type="password"
      value=""
    />
    <div
      class="c5 c6 c10"
    >
      <svg
        aria-label="Show password"
        class="c7"
        fill="currentColor"
        focusable="false"
        role="img"
        viewBox="0 0 16 16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Show password
        </title>
        <path
          d="M16 8s-3-5.5-8-5.5S0 8 0 8s3 5.5 8 5.5S16 8 16 8zM1.173 8a13.133 13.133 0 011.66-2.043C4.12 4.668 5.88 3.5 8 3.5c2.12 0 3.879 1.168 5.168 2.457A13.133 13.133 0 0114.828 8c-.058.087-.122.183-.195.288-.335.48-.83 1.12-1.465 1.755C11.879 11.332 10.119 12.5 8 12.5c-2.12 0-3.879-1.168-5.168-2.457A13.134 13.134 0 011.172 8z"
        />
        <path
          d="M8 5.5a2.5 2.5 0 100 5 2.5 2.5 0 000-5zM4.5 8a3.5 3.5 0 117 0 3.5 3.5 0 01-7 0z"
        />
      </svg>
    </div>
  </div>
  <span
    class="c11"
  >
    Error message
  </span>
</div>
`
